packages/cinterop/src/jvmMain/kotlin/io/realm/jvm/SoLoader.kt

```diff
@@ -0,0 +1,197 @@
+ */
+
+package io.realm.jvm
+
+import java.io.File
+import java.nio.file.Files
+import java.security.MessageDigest
+import java.util.Collections
+import java.util.Enumeration
+import java.util.LinkedList
+import java.util.Locale
+import java.util.Properties
+
+/**
+ * Load the C++ dynamic libraries from the fat Jar.
+ * The fat Jar contains three platforms (Win, Linux and Mac) the loader detects the host platform
+ * then extract and install the libraries in the same order specified in the 'dynamic_libraries.properties' file.
+ *
+ * Note: this class should be invoke dynamically using reflection so the classloader can have accesses
+ * to the dynamic libraries files located inside the fat Jar.
+ */
+class SoLoader {
+    private val platform: Platform = Platform.currentOS()
+    private val libs: MutableList<Pair<String, String>> = mutableListOf()
+
+    init {
+        readLibrariesHashes()
+    }
+
+    fun load() {
+        // load the libraries in the order of dependency specified in 'dynamic_libraries.properties'
+        for (lib in libs) {
+            load(libraryName = lib.first, expectedHash = lib.second)
+        }
+    }
+
+    private fun load(libraryName: String, expectedHash: String) {
+        // load the embedded .so file located inside the Jar file.
+        // unpacking the file is skipped if the hash of the file is already installed.
+        // instead, the on-disk file will be loaded.
+
+        // for each SO file:
+        // check if the library is already installed in the default platform location
+        // path should be <default user lib dir>/io.realm.kotlin/hash/librealmffi.so
+        // if the full path exists (and the on-disk hash matches) then load it otherwise unpack and load it.
+        val libraryInstallationLocation: File = defaultAbsolutePath(libraryName, expectedHash)
+        if (!libraryInstallationLocation.exists()) {
+            unpackAndInstall(libraryName, libraryInstallationLocation, expectedHash)
+        } else {
+            // only double check the installed lib hash (in case it was tampered with locally)
+            validHashOrThrow(libraryInstallationLocation, expectedHash)
+        }
+        @Suppress("UnsafeDynamicallyLoadedCode")
+        // System.loadLibrary does not accept a full path to the lib (needs to be in the current Java paths)
+        System.load(libraryInstallationLocation.absolutePath)
+    }
+
+    private fun readLibrariesHashes() {
+        javaClass.getResourceAsStream("${platform.shortName}/dynamic_libraries.properties").use { props ->
+            OrderedProperties().run {
+                load(props)
+                for (libName in keys()) {
+                    libs.add(Pair(libName as String, get(libName) as String))
+                }
+            }
+        }
+    }
+
+    private fun defaultAbsolutePath(libraryName: String, libraryHash: String): File {
+        return File(
+            platform.defaultSystemLocation + File.separator +
```

Comment:
Would it make sense to be able to override the `platform.defaultSystemLocation` by an environment variable, so that people could put it somewhere else if they don't like our convention?

Reply:
Yes good point, I'll create an issue 👍